src/util/ThreadSafeQueue.h

```diff
@@ -17,18 +18,21 @@ class ThreadSafeQueue {
   std::mutex _mutex;
   std::condition_variable _pushNotification;
   std::condition_variable _popNotification;
-  bool _finished = false;
-  bool _aborted = false;
+  bool _lastElementPushed = false;
+  bool _disablePush = false;
   size_t _maxSize;
 
  public:
   explicit ThreadSafeQueue(size_t maxSize) : _maxSize{maxSize} {}
 
+  /// Pushes an element into the queue and blocks if the queue's capacity
+  /// is currently exhausted until some space is freed via pop() or abort()
+  /// is called to cancel execution

```

Comment:
Push...  . Block until there is free space in the queue or until `disablePush` was called. Return false iff `disablePush` was called. In this case the current  element and all future elements are not added to the queue.